File: cplusplus/322CoinChange.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> v(amount + 1, 0);
        v[0] = 0;
        for (int i = 1; i <= amount; ++i) {
            v[i] = INT_MAX;
            for (int j = 0; j < coins.size(); ++j) {
                if (i - coins[j] >= 0) {
                    if (v[i - coins[j]] != INT_MAX && v[i] > v[i - coins[j]] + 1) {
                        v[i] = v[i - coins[j]] + 1;
                    }
                }
            }
        }
        
        return v[amount] == INT_MAX? -1 : v[amount];
    }
};

File: java/322CoinChange.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (coins == null || coins.length == 0 || amount == 0)
            return 0;
        
        int[] changes = new int[amount + 1];
        for (int i = 1; i <= amount; ++i)
            changes[i] = Integer.MAX_VALUE;
        
        Arrays.sort(coins);
        for (int c: coins)
            for (int i = 1; i <= amount; ++i) {
                if (i >= c && changes[i - c] != Integer.MAX_VALUE) {
                    int newChange = changes[i - c] + 1;
                    if (newChange < changes[i])
                        changes[i] = newChange;
                }
            }
        
        if (changes[amount] == Integer.MAX_VALUE)
            return -1;
        
        return changes[amount];
    }
}
